src/node_report.cc

```diff
@@ -0,0 +1,1115 @@
+    outstream = &std::cerr;
+  } else {
+    // Regular file. Append filename to directory path if one was specified
+    if (strlen(report_directory) > 0) {
+      char pathname[NR_MAXPATH + NR_MAXNAME + 1] = "";
+#ifdef _WIN32
+      snprintf(pathname, sizeof(pathname),
+               "%s%s%s", report_directory, "\\", filename);
+#else
+      snprintf(pathname, sizeof(pathname),
+               "%s%s%s", report_directory, "/", filename);
+#endif
+      outfile.open(pathname, std::ios::out);
+    } else {
+      outfile.open(filename, std::ios::out);
+    }
+    // Check for errors on the file open
+    if (!outfile.is_open()) {
+      if (strlen(report_directory) > 0) {
+        std::cerr << "\nFailed to open Node.js report file: "
+                  << filename << " directory: " << report_directory
+                  << " (errno: " << errno << ")\n";
+      } else {
+        std::cerr << "\nFailed to open Node.js report file: "
+                  << filename << " (errno: " << errno << ")\n";
+      }
+      return;
+    } else {
+      std::cerr << "\nWriting Node.js report to file: " << filename << "\n";
+    }
+  }
+
+  // Pass our stream about by reference, not by copying it.
+  std::ostream &out = outfile.is_open() ? outfile : *outstream;
+
+  WriteNodeReport(isolate, event, message,
+                  location, filename, out, error, &tm_struct);
+
+  // Do not close stdout/stderr, only close files we opened.
+  if (outfile.is_open()) {
+    outfile.close();
+  }
+
+  std::cerr << "Node.js report completed\n";
+  if (name != nullptr) {
+    snprintf(name, NR_MAXNAME + 1, "%s", filename);  // return report file name
+  }
+}
+
+/*****************************************************************************
+ * External function to trigger a node report, writing to a supplied stream.
+ *
+ *****************************************************************************/
+void GetNodeReport(Isolate* isolate, DumpEvent event, const char* message,
+                   const char* location, MaybeLocal<Value> error,
+                   std::ostream& out) {
+  // Obtain the current time and the pid (platform dependent)
+  TIME_TYPE tm_struct;
+#ifdef _WIN32
+  GetLocalTime(&tm_struct);
+#else  // UNIX, OSX
+  struct timeval time_val;
+  gettimeofday(&time_val, nullptr);
+  localtime_r(&time_val.tv_sec, &tm_struct);
+#endif
+  WriteNodeReport(isolate, event, message,
+                  location, nullptr, out, error, &tm_struct);
+}
+
+/*******************************************************************************
+ * Internal function to coordinate and write the various sections of the node
+ * report to the supplied stream
+ *******************************************************************************/
+static void WriteNodeReport(Isolate* isolate, DumpEvent event,
+                            const char* message, const char* location,
+                            char* filename, std::ostream &out,
```

Comment:
style nit: left-leaning references

You might get through a lot of these just by running `make format-cpp` (possibly preceded by `make format-cpp-build`). There’s also https://github.com/nodejs/node/blob/master/CPP_STYLE_GUIDE.md.

Reply:
done. (thanks for the pointers)